Update rook to latest master build.

obnoxxx opened the pull request:
For 4.8 preparation, we need a newer rook image.

obnoxxx commented:
rook update pulls in an update of the controller-runtime dependency from 0.6.3 to 0.8.0.
Seems we need to adapt our code to that.

obnoxxx commented:
Weeding out the build errors one by one...
I am left with two:

```
 make
Running deps-update
go mod tidy && go mod vendor
go: creating new go.mod: module tmp
go get: added sigs.k8s.io/controller-tools v0.4.1
Updating generated code
/Users/madam/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
Building the ocs-operator binary
hack/go-build.sh
# github.com/openshift/ocs-operator/controllers/storagecluster
controllers/storagecluster/external_resources.go:182:4: cannot use "60s" (type string) as type *"k8s.io/apimachinery/pkg/apis/meta/v1".Duration in field value
controllers/storagecluster/kms_resources.go:79:38: cannot use runtimeObj (type "k8s.io/apimachinery/pkg/runtime".Object) as type client.Object in argument to r.Client.Delete:
	"k8s.io/apimachinery/pkg/runtime".Object does not implement client.Object (missing GetAnnotations method)
make: *** [build-go] Error 2
```

obnoxxx commented:
Remaining error:
```
# github.com/openshift/ocs-operator/controllers/storagecluster
controllers/storagecluster/kms_resources.go:79:38: cannot use runtimeObj (type "k8s.io/apimachinery/pkg/runtime".Object) as type client.Object in argument to r.Client.Delete:
	"k8s.io/apimachinery/pkg/runtime".Object does not implement client.Object (missing GetAnnotations method)
```

obnoxxx commented:
Locally, I am getting this error now:
```
"error loading manifests from directory: error checking provided apis in bundle ocs-operator.v4.8.0: couldn't find ceph.rook.io/v1/CephNFS (cephnfses) in bundle. found: map[noobaa.io/v1alpha1/BackingStore (backingstores):{} noobaa.io/v1alpha1/BucketClass (bucketclasses):{} noobaa.io/v1alpha1/NamespaceStore (namespacestores):{} noobaa.io/v1alpha1/NooBaa (noobaas):{} objectbucket.io/v1alpha1/ObjectBucket (objectbuckets):{} objectbucket.io/v1alpha1/ObjectBucketClaim (objectbucketclaims):{} ocs.openshift.io/v1/OCSInitialization (ocsinitializations):{} ocs.openshift.io/v1/StorageCluster (storageclusters):{}]"
```

leseb commented on controllers/storagecluster/external_resources.go:
```diff
@@ -176,10 +176,11 @@ func (r *StorageClusterReconciler) newExternalCephObjectStoreInstances(
 		return nil, err
 	}
 	// enable bucket healthcheck
+	time60s, _ := time.ParseDuration("60s")
```
You don't need to parse time anymore, just use `1 * time.Minute`

obnoxxx replied:
Thanks, fixed (yet to push). Actually `time.Minute` is sufficient.

leseb reviewed (commented):
Why did all the CRDs get deleted in https://github.com/openshift/ocs-operator/pull/1141/commits/c428b8bcf2041c38a7cc337741eb5e87f2cd4da7?

obnoxxx commented:
> Why did all the CRDs get deleted in [c428b8b](https://github.com/openshift/ocs-operator/commit/c428b8bcf2041c38a7cc337741eb5e87f2cd4da7)?

Right that is weird... Investigating.

obnoxxx commented:
updated with the current state
still some lint/test compile errors left
and then to clarify why the crds got removed after updating the rook version

obnoxxx commented:
We are making progress. Now rook's image is fixed upstream via https://github.com/rook/rook/pull/7512 and the image target starts to pass. Still some things to fix from the controller-runtime update...

obnoxxx commented:
@jarrpa - I incorporated your patches on top of my PR and rebased. Looks a lot better! 😄 Only waiting for the e2e tests to complete now - everything else passed!

obnoxxx commented:
From the test output:

```
...
time="2021-04-01T09:38:40Z" level=info msg="Pod ocs-operator-bundle-e2e-aws-ipi-install-install succeeded after 34m10s"
time="2021-04-01T09:38:40Z" level=info msg="Executing pod \"ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe\" running image \"stable:cli\""
time="2021-04-01T09:38:50Z" level=info msg="Container place-entrypoint in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe completed successfully"
time="2021-04-01T09:38:50Z" level=info msg="Container cp-entrypoint-wrapper in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe completed successfully"
time="2021-04-01T09:49:40Z" level=info msg="Container sidecar in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe completed successfully"
Error from server (NotFound): namespaces "openshift-storage" not found
OO_INSTALL_NAMESPACE is 'openshift-storage' which does not exist: creating
Installing "ocs-operator" in namespace "openshift-storage"
OO_TARGET_NAMESPACES is '!install': targeting operator installation namespace (openshift-storage)
OperatorGroup does not exist: creating it
OperatorGroup name is "oo-z6rqx"
Creating CatalogSource
CatalogSource name is "oo-dzq4l"
Creating Subscription
Subscription name is "oo-qjvv5"
Waiting for installPlan to be created
installplan.operators.coreos.com/install-vz8f9 patched
Install Plan approved
Waiting for ClusterServiceVersion to become ready...
Timed out waiting for csv to become ready
Dumping Namespace openshift-storage as /logs/artifacts/ns-openshift-storage.yaml
Dumping OperatorGroup oo-z6rqx as /logs/artifacts/og-oo-z6rqx.yaml
Dumping CatalogSource oo-dzq4l as /logs/artifacts/cs-oo-dzq4l.yaml
Dumping Subscription oo-qjvv5 as /logs/artifacts/sub-oo-qjvv5.yaml
  Subscription oo-qjvv5 status state: UpgradePending
ClusterServiceVersion was never created
Dumping all ClusterServiceVersions in namespace openshift-storage to /logs/artifacts/openshift-storage-all-csvs.yaml
Dumping all installPlans in namespace openshift-storage as /logs/artifacts/installPlans.yaml
{"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-04-01T09:49:38Z"}
error: failed to execute wrapped command: exit status 1
time="2021-04-01T09:49:40Z" level=info msg="Container test in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe failed, exit code 1, reason Error"
time="2021-04-01T09:49:40Z" level=info msg="Executing pod \"ocs-operator-bundle-e2e-aws-gather-aws-console\" running image \"pipeline:ocp-4.5-upi-installer\""
time="2021-04-01T09:49:51Z" level=info msg="Container place-entrypoint in pod ocs-operator-bundle-e2e-aws-gather-aws-console completed successfully"
time="2021-04-01T09:49:51Z" level=info msg="Container cp-entrypoint-wrapper in pod ocs-operator-bundle-e2e-aws-gather-aws-console completed successfully"
...
```

Not sure if this is a flake or real.

FWIW, I see `pipeline:ocp-4.5-upi-installer` - I think we shouldn't be running OCP 4.5 for testing OCS 4.8 / master, right?

umangachapagain commented:
> From the test output:
> 
> ```
> ...
> time="2021-04-01T09:38:40Z" level=info msg="Pod ocs-operator-bundle-e2e-aws-ipi-install-install succeeded after 34m10s"
> time="2021-04-01T09:38:40Z" level=info msg="Executing pod \"ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe\" running image \"stable:cli\""
> time="2021-04-01T09:38:50Z" level=info msg="Container place-entrypoint in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe completed successfully"
> time="2021-04-01T09:38:50Z" level=info msg="Container cp-entrypoint-wrapper in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe completed successfully"
> time="2021-04-01T09:49:40Z" level=info msg="Container sidecar in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe completed successfully"
> Error from server (NotFound): namespaces "openshift-storage" not found
> OO_INSTALL_NAMESPACE is 'openshift-storage' which does not exist: creating
> Installing "ocs-operator" in namespace "openshift-storage"
> OO_TARGET_NAMESPACES is '!install': targeting operator installation namespace (openshift-storage)
> OperatorGroup does not exist: creating it
> OperatorGroup name is "oo-z6rqx"
> Creating CatalogSource
> CatalogSource name is "oo-dzq4l"
> Creating Subscription
> Subscription name is "oo-qjvv5"
> Waiting for installPlan to be created
> installplan.operators.coreos.com/install-vz8f9 patched
> Install Plan approved
> Waiting for ClusterServiceVersion to become ready...
> Timed out waiting for csv to become ready
> Dumping Namespace openshift-storage as /logs/artifacts/ns-openshift-storage.yaml
> Dumping OperatorGroup oo-z6rqx as /logs/artifacts/og-oo-z6rqx.yaml
> Dumping CatalogSource oo-dzq4l as /logs/artifacts/cs-oo-dzq4l.yaml
> Dumping Subscription oo-qjvv5 as /logs/artifacts/sub-oo-qjvv5.yaml
>   Subscription oo-qjvv5 status state: UpgradePending
> ClusterServiceVersion was never created
> Dumping all ClusterServiceVersions in namespace openshift-storage to /logs/artifacts/openshift-storage-all-csvs.yaml
> Dumping all installPlans in namespace openshift-storage as /logs/artifacts/installPlans.yaml
> {"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-04-01T09:49:38Z"}
> error: failed to execute wrapped command: exit status 1
> time="2021-04-01T09:49:40Z" level=info msg="Container test in pod ocs-operator-bundle-e2e-aws-optional-operators-ci-subscribe failed, exit code 1, reason Error"
> time="2021-04-01T09:49:40Z" level=info msg="Executing pod \"ocs-operator-bundle-e2e-aws-gather-aws-console\" running image \"pipeline:ocp-4.5-upi-installer\""
> time="2021-04-01T09:49:51Z" level=info msg="Container place-entrypoint in pod ocs-operator-bundle-e2e-aws-gather-aws-console completed successfully"
> time="2021-04-01T09:49:51Z" level=info msg="Container cp-entrypoint-wrapper in pod ocs-operator-bundle-e2e-aws-gather-aws-console completed successfully"
> ...
> ```
> 
> Not sure if this is a flake or real.
> 
> FWIW, I see `pipeline:ocp-4.5-upi-installer` - I think we shouldn't be running OCP 4.5 for testing OCS 4.8 / master, right?

We're using OCP 4.7.
The CSV was never created, so it doesn't look like a flake to me.

umangachapagain commented:
/retest

jarrpa commented:
Okay, the e2e failures may indeed be a flake. It seems it failed to even install the OCS CSV. Looking around, I found that the CatalogSource they generate to hold the CSV was in TRANSIENT_FAILURE:
```
time="2021-04-01T12:41:49Z" level=info msg="state.Key.Namespace=openshift-storage state.Key.Name=oo-bfs6t state.State=TRANSIENT_FAILURE"
time="2021-04-01T12:41:49Z" level=error msg="failed to list bundles: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 172.30.74.197:50051: connect: no route to host\"" catalog="{oo-bfs6t openshift-storage}"
```
So, as is often the case, we are probably dealing with network issues and/or API limits.

/test ocs-operator-bundle-e2e-aws

jarrpa commented:
/retest

obnoxxx commented:
Created PR #1144 for only the CSV part of the rook upgrade.
Interestingly, it also fails e2e tests, similarly to this one.

obnoxxx commented:
rebased and updated to latest rook update corresponding to #1150

obnoxxx commented:
added DNM label. if tests pass, the patches need some polishing

obnoxxx commented:
/test ocs-operator-ci

obnoxxx commented:
seems the cluster (OCS/ceph) didn't come up in the test

obnoxxx commented:
Not sure if this is the problem:

```
  - lastTransitionTime: "2021-04-09T08:55:02Z"
    lastUpdateTime: "2021-04-09T08:55:02Z"
    message: csv created in namespace with multiple operatorgroups, can't pick one automatically
    phase: Failed
    reason: TooManyOperatorGroups
  lastTransitionTime: "2021-04-09T08:55:02Z"
  lastUpdateTime: "2021-04-09T08:55:02Z"
  message: csv created in namespace with multiple operatorgroups, can't pick one automatically
  phase: Failed
  reason: TooManyOperatorGroups
```

found in https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_ocs-operator/1141/pull-ci-openshift-ocs-operator-master-ocs-operator-bundle-e2e-aws/1380431051205120000/artifacts/ocs-operator-bundle-e2e-aws/ocs-must-gather/artifacts/ocs-must-gather/registry-build02-ci-openshift-org-ci-op-rig6vt5p-stable-sha256-ad1489e3d7d201d7a9cc908a56ef6e58ec02ec871ae18aa83c263b092d49d49d/namespaces/openshift-storage/operators.coreos.com/clusterserviceversions/ocs-operator.v4.8.0.yaml

obnoxxx commented:
/test ocs-operator-bundle-e2e-aws

obnoxxx commented:
same problem as before

raghavendra-talur commented:
The latest test failures seem to be because of insufficient permissions for the ocs-operator. This could be due to the update in the controller runtime.

The error message is 
```
2021-04-16T18:58:02.029785222Z E0416 18:58:02.029733       1 leaderelection.go:325] error retrieving resource lock openshift-storage/ab76f4c9.openshift.io: leases.coordination.k8s.io "ab76f4c9.openshift.io" is forbidden: User "system:serviceaccount:openshift-storage:ocs-operator" cannot get resource "leases" in API group "coordination.k8s.io" in the namespace "openshift-storage"
```

Looking at the recent fixes in the kubebuilder repo I see https://github.com/kubernetes-sigs/kubebuilder/pull/1809 which seems to indicate that the kubebuilder wasn't adding the required permissions to the operator. It might be fixed in the next release of the kubebuilder but until then we can probably add the permissions manually. 

I have added a commit(https://github.com/openshift/ocs-operator/pull/1152/commits/7852a8146c1f3cf45bd6fe8b6b00c312dc2b0e51) in my test PR which adds the required permissions and is also rebased on top of this PR. If it passes the tests please pick that commit.

obnoxxx commented:
@raghavendra-talur yay, your patch fixes it!
I'm going to polish this PR and then we can move forward.

obnoxxx commented:
When working on polishing / squashing the patches, I came across the test code that @jarrpa  commented out. Removing the comments lets the unit tests fail this way:

```
=== RUN   TestCreateSCCs                                                                                                                              
    ocsinitialization_controller_test.go:234:                                                                                                         
                Error Trace:    ocsinitialization_controller_test.go:234                                                                              
                Error:          Received unexpected error:                                                                                            
                                Operation cannot be fulfilled on ocsinitializations.ocs.openshift.io "ocsinit": object was modified                   
                Test:           TestCreateSCCs                                                                                                        
                Messages:       [Case 1]: failed to update ocsInit status                                                                             
--- FAIL: TestCreateSCCs (0.00s)                                                                                                                      
```

Not sure really how to fix this. But ideally, we should be commenting out such lines...

obnoxxx commented:
@jarrpa , @raghavendra-talur , @umangachapagain  - any idea?

jarrpa commented:
> When working on polishing / squashing the patches, I came across the test code that @jarrpa commented out. Removing the comments lets the unit tests fail this way:
> 
> ```
> === RUN   TestCreateSCCs                                                                                                                              
>     ocsinitialization_controller_test.go:234:                                                                                                         
>                 Error Trace:    ocsinitialization_controller_test.go:234                                                                              
>                 Error:          Received unexpected error:                                                                                            
>                                 Operation cannot be fulfilled on ocsinitializations.ocs.openshift.io "ocsinit": object was modified                   
>                 Test:           TestCreateSCCs                                                                                                        
>                 Messages:       [Case 1]: failed to update ocsInit status                                                                             
> --- FAIL: TestCreateSCCs (0.00s)                                                                                                                      
> ```
> 
> Not sure really how to fix this. But ideally, we should be commenting out such lines...

No idea. I wasn't able to figure anything further out, hence why I had to leave it like that just to make progress. It has something to do with how the fake reconciler is implemented, but that is still fairly opaque to me.

obnoxxx commented:
@jarrpa - do you suggest to move forward with the commented-out lines? 

(It just feels odd, since the lines seemed to have been there for a purpose... ;-)

jarrpa commented:
For the sake of expediency, yes. There should at least me a TODO to figure out why they were suddenly a problem.

obnoxxx commented:
updated to the latest and greatest rook image rook/ceph:v1.6.0.95.gf4cfc7a
and polished the patches

still the only weird thing is the commented out test code lines... I added a TODO comment

jarrpa commented:
/override ci/prow/ci-index-dev-master-dependencies
